File: frontend/src/lib/crypto.ts
// SharePwd.io — Burn After Reading
// Zero-knowledge secret sharing. Client-side AES-256-GCM encryption, secrets self-destruct after viewing.
// Copyright (c) 2025-2026 Antonin HILY — CTO, Jizo AI

/**
 * Zero-knowledge encryption module using Web Crypto API.
 * All encryption/decryption happens client-side.
 * The server NEVER sees plaintext data.
 */

const PBKDF2_ITERATIONS = 600_000;
const AES_KEY_LENGTH = 256;
const IV_LENGTH = 12;
const SALT_LENGTH = 16;

export interface EncryptResult {
  encryptedData: string;
  iv: string;
  key: string;
}

export interface EncryptWithPassphraseResult {
  encryptedData: string;
  iv: string;
  salt: string;
}

function copyToArrayBuffer(src: Uint8Array): ArrayBuffer {
  const dst = new ArrayBuffer(src.byteLength);
  new Uint8Array(dst).set(src);
  return dst;
}

function toBase64(data: Uint8Array | ArrayBuffer): string {
  const bytes = data instanceof Uint8Array ? data : new Uint8Array(data);
  let binary = "";
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

function fromBase64(base64: string): Uint8Array {
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

function toBase64Url(data: Uint8Array | ArrayBuffer): string {
  return toBase64(data)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

function fromBase64Url(base64url: string): Uint8Array {
  let base64 = base64url.replace(/-/g, "+").replace(/_/g, "/");
  while (base64.length % 4) {
    base64 += "=";
  }
  return fromBase64(base64);
}

async function generateKey(): Promise<CryptoKey> {
  return crypto.subtle.generateKey(
    { name: "AES-GCM", length: AES_KEY_LENGTH },
    true,
    ["encrypt", "decrypt"]
  );
}

async function deriveKeyFromPassphrase(
  passphrase: string,
  salt: Uint8Array
): Promise<CryptoKey> {
  const encoder = new TextEncoder();
  const keyMaterial = await crypto.subtle.importKey(
    "raw",
    encoder.encode(passphrase),
    "PBKDF2",
    false,
    ["deriveKey"]
  );

  return crypto.subtle.deriveKey(
    {
      name: "PBKDF2",
      salt: copyToArrayBuffer(salt) as BufferSource,
      iterations: PBKDF2_ITERATIONS,
      hash: "SHA-256",
    },
    keyMaterial,
    { name: "AES-GCM", length: AES_KEY_LENGTH },
    false,
    ["encrypt", "decrypt"]
  );
}

export async function encryptText(plaintext: string): Promise<EncryptResult> {
  const key = await generateKey();
  const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
  const encoder = new TextEncoder();

  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: copyToArrayBuffer(iv) as BufferSource },
    key,
    encoder.encode(plaintext)
  );

  const rawKey = await crypto.subtle.exportKey("raw", key);

  return {
    encryptedData: toBase64(new Uint8Array(ciphertext)),
    iv: toBase64(iv),
    key: toBase64Url(new Uint8Array(rawKey)),
  };
}

export async function decryptText(
  encryptedData: string,
  iv: string,
  keyBase64Url: string
): Promise<string> {
  const keyBytes = fromBase64Url(keyBase64Url);
  const key = await crypto.subtle.importKey(
    "raw",
    copyToArrayBuffer(keyBytes),
    { name: "AES-GCM", length: AES_KEY_LENGTH },
    false,
    ["decrypt"]
  );

  const ivBytes = fromBase64(iv);
  const cipherBytes = fromBase64(encryptedData);

  const decrypted = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: copyToArrayBuffer(ivBytes) as BufferSource },
    key,
    copyToArrayBuffer(cipherBytes)
  );

  return new TextDecoder().decode(decrypted);
}

export async function encryptWithPassphrase(
  plaintext: string,
  passphrase: string
): Promise<EncryptWithPassphraseResult> {
  const salt = crypto.getRandomValues(new Uint8Array(SALT_LENGTH));
  const key = await deriveKeyFromPassphrase(passphrase, salt);
  const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
  const encoder = new TextEncoder();

  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: copyToArrayBuffer(iv) as BufferSource },
    key,
    encoder.encode(plaintext)
  );

  return {
    encryptedData: toBase64(new Uint8Array(ciphertext)),
    iv: toBase64(iv),
    salt: toBase64(salt),
  };
}

export async function decryptWithPassphrase(
  encryptedData: string,
  iv: string,
  salt: string,
  passphrase: string
): Promise<string> {
  const saltBytes = fromBase64(salt);
  const key = await deriveKeyFromPassphrase(passphrase, saltBytes);

  const ivBytes = fromBase64(iv);
  const cipherBytes = fromBase64(encryptedData);

  const decrypted = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: copyToArrayBuffer(ivBytes) as BufferSource },
    key,
    copyToArrayBuffer(cipherBytes)
  );

  return new TextDecoder().decode(decrypted);
}

export { toBase64, fromBase64, toBase64Url, fromBase64Url };

File: cli/crypto.go
// SharePwd.io — Burn After Reading
// Zero-knowledge secret sharing. Client-side AES-256-GCM encryption, secrets self-destruct after viewing.
// Copyright (c) 2025-2026 Antonin HILY — CTO, Jizo AI

package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"

	"golang.org/x/crypto/pbkdf2"
)

const (
	pbkdf2Iterations = 600_000
	aesKeyLength     = 32 // 256 bits
	ivLength         = 12 // 96 bits, standard GCM nonce
	saltLength       = 16 // 128 bits
)

// EncryptResult holds keyless encryption output.
type EncryptResult struct {
	EncryptedData string // Base64 standard
	IV            string // Base64 standard
	Key           string // Base64URL (no padding) — goes in URL fragment
}

// EncryptWithPassphraseResult holds passphrase-based encryption output.
type EncryptWithPassphraseResult struct {
	EncryptedData string // Base64 standard
	IV            string // Base64 standard
	Salt          string // Base64 standard
}

// EncryptKeyless encrypts plaintext with a random AES-256-GCM key.
// Compatible with frontend encryptText().
func EncryptKeyless(plaintext []byte) (*EncryptResult, error) {
	key := make([]byte, aesKeyLength)
	if _, err := io.ReadFull(rand.Reader, key); err != nil {
		return nil, fmt.Errorf("generate key: %w", err)
	}

	iv := make([]byte, ivLength)
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, fmt.Errorf("generate iv: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("aes cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("gcm: %w", err)
	}

	ciphertext := gcm.Seal(nil, iv, plaintext, nil)

	return &EncryptResult{
		EncryptedData: base64.StdEncoding.EncodeToString(ciphertext),
		IV:            base64.StdEncoding.EncodeToString(iv),
		Key:           base64.RawURLEncoding.EncodeToString(key),
	}, nil
}

// DecryptKeyless decrypts data encrypted with EncryptKeyless.
// Compatible with frontend decryptText().
func DecryptKeyless(encryptedDataB64, ivB64, keyB64URL string) ([]byte, error) {
	key, err := base64.RawURLEncoding.DecodeString(keyB64URL)
	if err != nil {
		return nil, fmt.Errorf("decode key: %w", err)
	}

	iv, err := base64.StdEncoding.DecodeString(ivB64)
	if err != nil {
		return nil, fmt.Errorf("decode iv: %w", err)
	}

	ciphertext, err := base64.StdEncoding.DecodeString(encryptedDataB64)
	if err != nil {
		return nil, fmt.Errorf("decode ciphertext: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("aes cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("gcm: %w", err)
	}

	plaintext, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}

	return plaintext, nil
}

// EncryptWithPassphrase encrypts plaintext using PBKDF2-derived key.
// Compatible with frontend encryptWithPassphrase().
func EncryptWithPassphrase(plaintext []byte, passphrase string) (*EncryptWithPassphraseResult, error) {
	salt := make([]byte, saltLength)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
		return nil, fmt.Errorf("generate salt: %w", err)
	}

	key := pbkdf2.Key([]byte(passphrase), salt, pbkdf2Iterations, aesKeyLength, sha256.New)

	iv := make([]byte, ivLength)
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return nil, fmt.Errorf("generate iv: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("aes cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("gcm: %w", err)
	}

	ciphertext := gcm.Seal(nil, iv, plaintext, nil)

	return &EncryptWithPassphraseResult{
		EncryptedData: base64.StdEncoding.EncodeToString(ciphertext),
		IV:            base64.StdEncoding.EncodeToString(iv),
		Salt:          base64.StdEncoding.EncodeToString(salt),
	}, nil
}

// DecryptWithPassphrase decrypts data encrypted with EncryptWithPassphrase.
// Compatible with frontend decryptWithPassphrase().
func DecryptWithPassphrase(encryptedDataB64, ivB64, saltB64, passphrase string) ([]byte, error) {
	salt, err := base64.StdEncoding.DecodeString(saltB64)
	if err != nil {
		return nil, fmt.Errorf("decode salt: %w", err)
	}

	key := pbkdf2.Key([]byte(passphrase), salt, pbkdf2Iterations, aesKeyLength, sha256.New)

	iv, err := base64.StdEncoding.DecodeString(ivB64)
	if err != nil {
		return nil, fmt.Errorf("decode iv: %w", err)
	}

	ciphertext, err := base64.StdEncoding.DecodeString(encryptedDataB64)
	if err != nil {
		return nil, fmt.Errorf("decode ciphertext: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("aes cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("gcm: %w", err)
	}

	plaintext, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}

	return plaintext, nil
}
